pkg/cloudprovider/provider/vsphere/provider.go

```diff
@@ -88,7 +98,75 @@ func (vsphereServer VSphereServer) Status() instance.Status {
 }
 
 func (p *provider) AddDefaults(spec v1alpha1.MachineSpec) (v1alpha1.MachineSpec, bool, error) {
-	return spec, false, nil
+	changed := false
+
+	cfg, _, rawCfg, err := p.getConfig(spec.ProviderConfig)
+	if err != nil {
+		return spec, changed, cloudprovidererrors.TerminalError{
+			Reason:  v1alpha1.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	// default templatenetname to first network of template if none specific was given.
+	if cfg.TemplateNetName == "" && cfg.VMNetName != "" {
+		ctx := context.TODO()
+		client, err := getClient(cfg.Username, cfg.Password, cfg.VSphereURL, cfg.AllowInsecure)
+		if err != nil {
+			return spec, changed, fmt.Errorf("failed to get vsphere client: '%v'", err)
+		}
+		defer client.Logout(ctx)
+
+		finder, err := getDatacenterFinder(cfg.Datacenter, client)
+		if err != nil {
+			return spec, changed, err
+		}
+
+		templateVm, err := finder.VirtualMachine(ctx, cfg.TemplateVMName)
+		if err != nil {
+			return spec, changed, err
+		}
+
+		availableData, err := getNetworkDevicesAndBackingsFromVM(ctx, templateVm, "")
+		if err != nil {
+			return spec, changed, err
+		}
+		if len(availableData) == 0 {
+			return spec, changed, errors.New("found no network adapter in template vm")
+		}
+
```

Comment:
please also check if `len(availableData) == 1`, otherwise we should not default. Also, please rename `availableData` to `availableNetworkDevices` so its a little easier to read :) Otherwise were good to go